File: app/src/androidTest/java/com/xmartlabs/template/ui/common/RecyclerViewInteraction.java
package com.xmartlabs.template.ui.common;

import android.support.test.espresso.NoMatchingViewException;
import android.view.View;

import org.hamcrest.Matcher;

import java.util.List;

import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.contrib.RecyclerViewActions.scrollToPosition;

/**
 * Created by medina on 21/09/2016.
 * Copied from https://medium.com/@_rpiel/recyclerview-and-espresso-a-complicated-story-3f6f4179652e#.dm9yzv52b.
 */
public class RecyclerViewInteraction<A> {
  private Matcher<View> viewMatcher;
  private List<A> items;

  private RecyclerViewInteraction(Matcher<View> viewMatcher) {
    this.viewMatcher = viewMatcher;
  }

  public static <A> RecyclerViewInteraction<A> onRecyclerView(Matcher<View> viewMatcher) {
    return new RecyclerViewInteraction<>(viewMatcher);
  }

  public RecyclerViewInteraction<A> withItems(List<A> items) {
    this.items = items;
    return this;
  }

  public RecyclerViewInteraction<A> check(ItemViewAssertion<A> itemViewAssertion) {
    for (int i = 0; i < items.size(); i++) {
      onView(viewMatcher)
          .perform(scrollToPosition(i))
          .check(new RecyclerItemViewAssertion<>(i, items.get(i), itemViewAssertion));
    }
    return this;
  }

  public interface ItemViewAssertion<A> {
    void check(A item, View view, NoMatchingViewException e);
  }
}

File: app/src/androidTest/java/com/xmartlabs/template/ui/common/RecyclerViewInteraction.kt
package com.xmartlabs.template.ui.common

import android.support.test.espresso.Espresso.onView
import android.support.test.espresso.NoMatchingViewException
import android.support.test.espresso.contrib.RecyclerViewActions.scrollToPosition
import android.support.v7.widget.RecyclerView
import android.view.View
import org.hamcrest.Matcher

/**
 * Copied from https://medium.com/@_rpiel/recyclerview-and-espresso-a-complicated-story-3f6f4179652e#.dm9yzv52b.
 */
class RecyclerViewInteraction<A> private constructor(private val viewMatcher: Matcher<View>) {
  companion object {
    fun <A> onRecyclerView(viewMatcher: Matcher<View>) = RecyclerViewInteraction<A>(viewMatcher)
  }
  
  private var items: List<A>? = null

  fun withItems(items: List<A>) = apply { this.items = items }

  fun check(itemViewAssertion: ItemViewAssertion<A>) = apply {
    items?.forEachIndexed { index, element ->
      onView(viewMatcher)
          .perform(scrollToPosition<RecyclerView.ViewHolder>(index))
          .check(RecyclerItemViewAssertion(index, element, itemViewAssertion))
    }
  }

  interface ItemViewAssertion<in A> {
    fun check(item: A, view: View, e: NoMatchingViewException)
  }
}
